### ISSUE TITLE
Tracking Issue for `num_as_ne_bytes` feature

### ISSUE THREAD

**Issue by hch12907** [rust-lang/rust#76976]
The feature gate for the issue is `#![feature(num_as_ne_bytes)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implementation (#76610)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

**Comment by SimonSapin** [comment]
Ad discussed starting in https://github.com/rust-lang/rust/issues/64464#issuecomment-705786778 I’m not convinced this feature carries its weight to be included in the standard library, given that `to_ne_bytes` also exists.

**Comment by mbartlett21** [comment]
`as_ne_bytes_mut` may be more useful.

**Comment by SimonSapin** [comment]
How so? Trying to go beyond "I want to borrow because I want to borrow"

**Comment by shepmaster** [comment]
In a similar vein, I was curious if the opposite existed, converting from `&(mut) [u8; 4]` to `&(mut) u32`:

```rust
let mut a = [1, 2, 3, 4];
let v: &mut u32 = u32::from_ne_bytes_mut(&mut a);
// Then potentially used with
let x = AtomicU32::from_mut(v);
```

I do not have a specific usecase in mind. My thoughts were likewise around avoiding the copy (small as it may be).

**Comment by m-ou-se** [comment]
@shepmaster Something kind of like that exists: [`[u8]::align_to::<u32>`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.align_to)

It returns `(&[u8], &[u32], &[u8])` because the byte slice might not be aligned properly for an `u32`, so it returns the unaligned parts separately.

It's also unsafe because it can transmute to any type. A safe version for integers could be possible, but that won't solve the alignment problem.

**Comment by shepmaster** [comment]
A good point! The original case had a `repr(align)`, which I neglected to show here. You get that "for free" if you store a `u32` and have something like the suggested `as_ne_bytes_mut`.

**Comment by hch12907** [comment]
Sorry for replying really late here (I accidentally unsubscribed from the issue.) but how should I proceed from here?

* Remove the function from std as per Simon's suggestion
* Add another function `as_ne_bytes_mut`

Though, the `AtomicU*` one does sound interesting to me although it is unrelated my original use case.

**Comment by sffc** [comment]
This feature would be useful to make this be a safe conversion.  Currently this requires unsafe code.  We use this type of abstraction in [TinyStr](https://github.com/zbraniecki/tinystr); it would also be handy in some serialization-type libraries.

**Comment by SimonSapin** [comment]
> to make this be a safe conversion

Which "this" are you referring to? Why does `to_ne_bytes` not work for you?

**Comment by sffc** [comment]
> > to make this be a safe conversion
> 
> Which "this" are you referring to? Why does `to_ne_bytes` not work for you?

I meant that `as_ne_bytes()` allows the conversion from `&u32` to `&[u8]` to be safe. Right now it requires `unsafe`.

**Comment by sffc** [comment]
Also: in conjunction with https://github.com/rust-lang/rust/issues/74985#issuecomment-840068608, this feature would also enable safe conversion from `&[u32]` to `&[u8]`.

**Comment by hch12907** [comment]
Regarding `Aligned<[u8; 4], 4>`: I believe that would be out of scope for this feature?

At least, I am not aware of any machinery in `std` that allows one to say "the returned value is N-byte aligned", so doing that would require a new feature that mimics the `aligned` crate in `std`.

**Comment by scottmcm** [comment]
I think I'm with https://github.com/rust-lang/rust/issues/76976#issuecomment-706764305 on this one.  Given that this can't (portably) have `.as_le_bytes()` and `.as_be_bytes()`, I'm not sure it's worth having as a distinct thing, assuming that the Safe Transmute project lands -- as one of its top two use cases is offering `&[u8]` native-endian views into everything where it's safe to do so, including `f32` and `u64` and such.

**Comment by BurntSushi** [comment]
@scottmcm Agreed. I think it would be great to wait and see what the Safe Transmute project gives us here.

**Comment by workingjubilee** [comment]
The use case is valid but a more general mechanism is direly needed here.

**Comment by hch12907** [comment]
Opened #85679 which removes this feature.
